### ISSUE TITLE
Make CanJS compatible with how webpack strips out dev code

### ISSUE THREAD

**Issue by chasenlehara** [canjs/canjs#4011]
__TLDR:__ use `process.env.NODE_ENV !== 'production'` in CanJS’s code to support how other bundlers (like webpack) remove code from production builds.

> This was discussed on a [recent live stream (34:36)](https://youtu.be/coSEECcyi00?t=34m36s) and a [previous live stream (18:28)](https://youtu.be/EEBytcIX7oA?t=18m28s).

## Overview

Right now, CanJS relies on steal’s `removeDevelopmentCode` [build option](https://stealjs.com/docs/steal-tools.BuildOptions.html) to remove code when built for production:

```js
//!steal-remove-start
  // Code removed from the production build
//!steal-remove-end
```

This doesn’t work out of the box with other bundlers, which support something like this ([Browserify](https://github.com/hughsk/envify#envify---), [webpack)](https://webpack.js.org/guides/production/#specify-the-environment):

```js
if (process.env.NODE_ENV !== 'production') {
  // Code removed from the production build
}
```

We can start using the above in our code today by using steal’s `envify`  [build option](https://stealjs.com/docs/steal-tools.BuildOptions.html).

Unfortunately, replacing our proprietary `//!steal-remove-*` comments with `process.env.NODE_ENV` checks would cause existing steal users to start having debug code in their production build, so until CanJS 5, I think we would need to have both in our source code like this:

```js
//!steal-remove-start
if (process.env.NODE_ENV !== 'production') {
  // Code removed from the production build
}
//!steal-remove-end
```

This is _potentially_ codemod-able, but because we sometimes use the steal comments inside objects, like this:

```js
funcCall({
  //!steal-remove-start
  debugArg: true
  //!steal-remove-end
})
```

…I think it would be better to just manually make the change everywhere. At the time of this writing, there are 139 [results for steal-remove-start in the CanJS org](https://github.com/search?q=org%3Acanjs+steal-remove-start&type=Code), so I think it would take about a week for someone to update all the code and release it.

## How we teach this

After we make this change, I think we should add a new “Production” topic to our guides that explains how to build for production with popular bundlers. Alternatively, we could have a more generic “Performance” page that includes this and other performance tips for CanJS as a whole.

Additionally, we should consider adding info about this and using `can-log` to our contribution guide, so contributors know how to correctly add useful debugging info that’s removed in production.

## Drawbacks & Alternatives

If this proposal is accepted “now” (Spring 2018), I think we should include both the steal comments and `process.env` checks, although this makes our code more difficult to maintain now and would suggest work in the future to remove the steal comments in the next major versions of each package.

Alternatively, we could wait to replace the steal comments with `process.env` checks in the next major version of each package, although that wouldn’t help packages like `can-util` in which we’re intentionally trying to avoid a new major version.

## Unresolved questions

Are there other alternatives or workarounds that I didn’t include in this proposal?

**Comment by matthewp** [comment]
You could wrap your conditionals with the comments :/

**Comment by r0m4n** [comment]
Does anyone know if there's any workaround for removing these dev warnings when bundling with something other than steal? For example, browserify

**Comment by cherifGsoul** [comment]
Here is a list to track changes in modules for the first part of the issue, [x] means is approved and merged:

- [x] can-ajax
- [x] can-assign
- [ ] can-attribute-encoder
- [ ] can-attribute-observable
- [x] can-bind
- [x] can-component
- [ ] can-compute
- [x] can-connect
- [ ] can-connect-feathers
- [ ] can-connect-tag
- [ ] can-construct
- [ ] can-construct-super
- [ ] can-control
- [ ] can-data-types
- [ ] can-debug
- [ ] can-define
- [ ] can-define-backup
- [ ] can-define-lazy-value
- [ ] can-define-stream
- [ ] can-define-stream-kefir
- [ ] can-define-validate-validatejs
- [ ] can-deparam
- [ ] can-diff
- [ ] can-dom-data
- [ ] can-dom-data-state
- [ ] can-dom-events
- [ ] can-dom-mutate
- [ ] can-event-dom-enter
- [ ] can-event-dom-radiochange
- [ ] can-event-queue
- [ ] can-fixture
- [ ] can-fixture-socket
- [ ] can-globals
- [ ] can-kefir
- [ ] can-key
- [ ] can-key-tree
- [ ] can-list
- [ ] can-local-store
- [ ] can-log
- [ ] can-make-map
- [ ] can-map
- [ ] can-map-define
- [ ] can-memory-store
- [ ] can-namespace
- [ ] can-ndjson-stream
- [ ] can-observation
- [ ] can-observation-recorder
- [x] can-observe
- [ ] can-param
- [ ] can-parse-uri
- [ ] can-query-logic
- [ ] can-queues
- [ ] can-realtime-rest-model
- [x] can-reflect
- [ ] can-reflect-dependencies
- [ ] can-reflect-promise
- [ ] can-reflect-tests
- [ ] can-rest-model
- [ ] can-route
- [ ] can-route-pushstate
- [ ] can-set-legacy
- [ ] can-simple-dom
- [ ] can-simple-map
- [ ] can-simple-observable
- [x] can-stache
- [ ] can-stache-bindings
- [ ] can-stache-converters
- [ ] can-stache-key
- [ ] can-stache-route-helpers
- [ ] can-stream
- [ ] can-stream-kefir
- [ ] can-string
- [ ] can-string-to-any
- [ ] can-super-model
- [ ] can-symbol
- [ ] can-test-helpers
- [ ] can-util
- [ ] can-validate
- [ ] can-validate-interface
- [ ] can-validate-legacy
- [ ] can-validate-validatejs
- [ ] can-vdom
- [ ] can-view-autorender
- [ ] can-view-callbacks
- [x] can-view-live
- [ ] can-view-model
- [ ] can-view-nodelist
- [ ] can-view-parser
- [ ] can-view-scope
- [ ] can-view-target
